### ISSUE TITLE
what's the difference between pulse and stylegan encoder?

### ISSUE THREAD

**Issue by danielkaifeng** [adamian98/pulse#29]
1. Stylegan encoder predicts face similar to input face, while pulse predicts face whose LR image close to input face.
2. Just like stylegan encoder, pulse can only predict face based on pretrained GAN like stylegan. If you want to predict something else, you must train a GAN first. In this aspect, it is not general fit SR model.

Thanks for correcting me if I miss anything.

**Comment by woctezuma** [comment]
I assume that what you call a "StyleGAN encoder" is code [like this one](https://github.com/rolux/stylegan2encoder). Similarly to what was then suggested [in the StyleGAN2 article](https://github.com/NVlabs/stylegan2#projecting-images-to-latent-space), the aim is to project a real image in the latent space of the generative model. So what you get is one of the closest generated images to your high-resolution input.

This allows to edit images by moving from their projection along specified latent directions. You will get results [like this one](https://github.com/tg-bomze/StyleGAN2-Face-Modificator), where Mona Lisa's pose is changed and then she is shown as a child.

The original paper is about training the StyleGAN2 generative model, and the ability to project real images is a nice property. People have then built tools to do so because they wanted to have fun editing real images.

![StyleGAN2 encoder and editor](https://i.imgur.com/71F1Xgs.jpg)

In contrast, PULSE looks for plausible images in the latent space which would downscale correctly with respect to the low-resolution image input, e.g. PULSE produces a plausible 1024x1024 image which would downscale close to a 16x16 image input. There are many potential 1024x1024 images, so what you get is one among very many suitable candidates.

![PULSE](https://i.imgur.com/9n1PMQO.png)

**Comment by danielkaifeng** [comment]
With pretrained projector's weight and tricky finetune skill, you can get very similar HR image, like this one:
![face1_01](https://user-images.githubusercontent.com/18606140/85965079-450c2400-b9ee-11ea-8c14-73ff04b65db0.png)
![face1_01](https://user-images.githubusercontent.com/18606140/85965154-784eb300-b9ee-11ea-953f-abbb12dc8a67.png)

For very blur face, you can use a very small learning rate and increase GAN loss to ensure it doesn't generate blur HR image.
![image](https://user-images.githubusercontent.com/18606140/85965352-fc089f80-b9ee-11ea-8325-3ea358e9f9da.png)
![1_01](https://user-images.githubusercontent.com/18606140/85965246-a7652480-b9ee-11ea-8f72-b8b1e52a3a1a.png)

I believe stylegan-encoder's projection method can do a good job in face SR, and PULSE improve it's generalization by changing the optimization target from (input vs predict) to (input vs predict's downsampling low-resolution image).

**Comment by danielkaifeng** [comment]
To make a step forward, I think both stylegan-encoder and PULSE couldn't make a general SR application. You can't get a full body person SR unless you train it. So the SR effect rely a lot on your pretrained GAN. You trained a good GAN, you get nice result and vice verse.

Even in face generation, we must align the face first, which means the eyes, mouth etc should in the absolute exact position. It would be much better to generalize it to less preprocess image input.

**Comment by yuqiu1233** [comment]
> With pretrained projector's weight and tricky finetune skill, you can get very similar HR image, like this one:
> ![face1_01](https://user-images.githubusercontent.com/18606140/85965079-450c2400-b9ee-11ea-8c14-73ff04b65db0.png)
> ![face1_01](https://user-images.githubusercontent.com/18606140/85965154-784eb300-b9ee-11ea-953f-abbb12dc8a67.png)
> 
> For very blur face, you can use a very small learning rate and increase GAN loss to ensure it doesn't generate blur HR image.
> ![image](https://user-images.githubusercontent.com/18606140/85965352-fc089f80-b9ee-11ea-8325-3ea358e9f9da.png)
> ![1_01](https://user-images.githubusercontent.com/18606140/85965246-a7652480-b9ee-11ea-8f72-b8b1e52a3a1a.png)
> 
> I believe stylegan-encoder's projection method can do a good job in face SR, and PULSE improve it's generalization by changing the optimization target from (input vs predict) to (input vs predict's downsampling low-resolution image).

Can you tell me how to finetune this project?  thanks
